keps/sig-testing/1746-stage-e2e-test-framework/README.md

```diff
@@ -0,0 +1,610 @@
+The SIG Testing sub-project
+[Testing Commons](https://github.com/kubernetes/community/tree/master/sig-testing#testing-commons)
+has been looking at ways of improving the user experience for contributors to
+the Kubernetes community.
+This KEP is part of that goal.
+
+We foresee the following benefits coming out of this work:
+
+1. Maintainability: the e2e test framework lacks a general API - there has been
+  little discussion on what parts of the framework do what.
+  From this work we plan to organize the code that is present and use only
+  external libraries and APIs available for e2e tests.
+  This decoupling will ensure the e2e test framework packages are better
+  contained allowing for e2e tests to have a more standard flow.
+2. External consumption: The Kubernetes ecosystem has grown considerably since
+  the creation of the e2e test framework. Some projects, such as the cloud
+  providers and CSI extensions, began within the core Kubernetes codebase and
+  have been migrated out or are in the process of being migrated out. We want
+  to provide them with the same set level of functionality they had in core
+  without having to deal with the problems of importing the entirety of core
+  Kubernetes.
+
+There are two main components to the e2e test framework: its configuration
+(used for setting up and running e2e tests) and its set of packages and utility
+functions.
+For this work, we plan on focusing on the latter.
+The entrypoint into the framework, the configuration part, will stay as is.
+Only the utility functions will be refactored.
+What this means for consumers, is that the functions offered via the e2e test
+framework will change but not the way e2e tests are setup and executed.
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+* Decouple the e2e test framework and all of its subpackages from any
+  Kubernetes internal packages (i.e., pkg, other e2e packages).
+* Impose import restrictions to prevent the use of core Kubernetes packages.
+* Allow the e2e test framework to be easily used and without requiring the
+  entirety of Kubernetes to be imported.
+* Refactor the e2e test framework so that its sub-packages are well contained
+  (i.e., `framework/pod/*` contains all necessary utilities to manage pods
+  during e2e tests).
+  * Update Kubernetes e2e tests to use the refactored version of any utility
+    function we change.
+
+### Non-Goals
+
+* Change the way e2e tests/jobs are setup or executed.
+* This enhancement is not directed at making the e2e test framework
+  provider-agnostic. Cloud provider functionality will be kept as is unless it
+  is used in inadequate e2e test framework sub-packages.
+* The e2e test framework will be refactored as needed strictly to remove its
+  dependence from other internal Kubernetes packages. It is not the goal of
+  this enhancement to propose an entire redesign of the e2e test framework.
```

Comment:
are we then saying that the organic design of this framework is suitable for export to 3p consumers?

Reply:
What i meant by this line is that, we are only focusing on decoupling the framework from core Kubernetes.
Any changes in the way the framework is designed will be a matter for a future discussion/KEP.
The assumption is that people are using the framework right now so we just want to improve their experience with that they are using as is.

There is another issue in k/kfor actually refactoring/redesigning the e2e test framework but we , testing commons, figure it would make more sense to clean up this part of Kubernetes before attempting that.

Reply:
Aaaah added some additional info here.
Reading the comment below made me realize the assumptions I was making while writing this part.
This bullet point is to indicate that the test context (the core of the framework), https://github.com/kubernetes/kubernetes/blob/0024c837bae4945427d790fdcd11176fb32aa056/test/e2e/framework/test_context.go#L73 , the "configuration" part of the e2e framework will be kept as is.